### ISSUE TITLE
Graph API structure tasks

### ISSUE THREAD

**Issue by dmlloyd** [qbicc/qbicc#70]
The graph API has evolved considerably over the period of this project.  After #67 it should be considered whether the graph API should be extracted (again) into a separate package or even module.  Included tasks are:

- [x] Examine the possibility of establishing a separate interface hierarchy for leaf node types, so that every possible node type is easily enumerable and visitable without confusing sub/supertype relationships (e.g. `TryInvocation` extends `Goto` because the normal path works like `Goto` does; should there be a separate interface for `Goto`-the-behavior versus `Goto`-the-leaf-node-type?)
- [x] Determine whether it is beneficial to have distinct leaf types for every `Value` type, even types with similar shape and properties (e.g. `CommutativeBinaryVakue` vs `Add`/`Multiply`/`CmpEq`/`Xor`/etc.)
- [x] Determine whether the leaf node types should be interfaces with implementation classes, or implementation classes directly (complicated by the multiply-inherited node traits defined by the `Value` hierarchy)
- [x] Crystalize the behavioral properties of the graph nodes, and strip away implementation which violates those properties:
    - `BasicBlock`s consist of two parts: the "label" (presently `NodeHandle`) and the terminator (i.e. `Terminal` instruction); since basic blocks form a control flow graph that may contain cycles, they necessarily must be allowed to be built thusly (in two stages)
    - Most nodes are not replaceable at any time, thus this functionality on `Node` is probably an attractive nuisance
    - `Type` ought not be a `Node` at all (this was an early design mistake when it was assumed that types would need to be substitutable, but this approach is not practical)
    - `BasicBlock` only extends `Node` *for* `NodeHandle`, and it is the only type that *uses* it; this should be cleaned up
- [x] Clean up extraneous node types:
    - ~~`PinnedNode` is an unnecessary intermediate, since `PhiValue` is the only possible subtype~~ I've introduced a "block entry" node which is pinned to the block, and on which the block's terminator must depend; the optimizer will be tasked with moving dependencies around in order to hoist dependants of the pinned node (this was already a reality though it was hidden from view due to using `null` as a placeholder to mean `block entry`)
    - `ValueTerminator` is no longer needed (it was needed when `GraphFactory` returned the node instead of the terminated basic block for invoke-in-try nodes)

This list is not exhaustive.
